### ISSUE TITLE
T10.59. Чи бувають віртуальні деструктори? Чому?

### ISSUE THREAD

**Issue by yanaskyrda** [kzhereb/knu-is-ooop2019#59]

**Comment by WAG13** [comment]
Так
Зазвичай віруальний деструктор оголошується тільки в базовому класі, в усіх успадкованих класах деструктор буде віртуальним за замовчуванням. Віртуальний деструктор гарантує, що всі необхідні деструктори для об'єкта такого класу будуть викликані, яким би чином ти не викликав для нього delete.

**Comment by AntonHladkiy** [comment]
Так , такий деструктор задається в базовому класі.такий деструктор гарантує що відбудеться видалення не тільки базового класу а і наслідуваних.

**Comment by MarinaMontag** [comment]
Віртуальні функції (і деструктор зокрема) в С ++ потрібні для того, щоб викликати перевантажений в класі-нащадку варіант відповідної функції через вказівник на базовий клас. Стосовно деструкторами це означає, що об'єкт класу-нащадка D можна коректно видалити, маючи покажчик на базовий клас B:
class B 
{
public:
virtual ~B() {}
};
...
class D : public B
{
public:
~D() {}
}
...
B * pObj = new D;
...
delete pObj; // - як би не virtual ~B(), ~D() не визвався б.

**Comment by RostyslavMV** [comment]
Так, віртуальні деструктори задаються в базовому класі, вони гарантують, що відбудеться видалення не тільки базового класу, а й усіх від нього наслідуваних.

**Comment by i1red** [comment]
Так. Якщо деструктор базового класу не є virtual, то за умови роботи через покажчик на базовий клас при знищенні об'єкта буде викликано тільки деструктор базового класу. virtual вирішує цю проблему

**Comment by tanyakurdelchuk** [comment]
Якщо базовий клас призначений для поліморфного використання, то його деструктор повинен оголошуватися віртуальним.  Для реалізації механізму віртуальних функцій кожен об'єкт класу зберігає вказівник на таблицю віртуальних функцій vptr, що збільшує його загальний розмір.  Зазвичай, при оголошенні віртуального деструктора такий клас вже має віртуальні функції, і збільшення розміру відповідного об'єкта не відбувається. Якщо ж базовий клас не призначений для поліморфного використання (не містить віртуальних функцій), то його деструктор не повинен оголошуватися віртуальним.

**Comment by ilyakorenevskyi** [comment]
Так ,без virtual не буде визвано деструктори классів наслідників

**Comment by Beam-of-Light** [comment]
Бувають, зазвичай тільки в базовому класі. За допомогою нього гарантовано викликаються деструктори класів наслідників.
